docs/csharp/nullable-attributes.md

```diff
@@ -0,0 +1,302 @@
+
+```csharp
+string GetTopLevelDomainFromFullUrl(string url);
+```
+
+If the `url` argument isn't null, the output isn't `null`. Once nullable references are enabled, that signature works correctly, provided your API never accepts a null input. However, if the input could be null, then then return value could also be null. Therefore, you could change the signature to the following:
+
+```csharp
+string? GetTopLevelDomainFromFullUrl(string? url);
+```
+
+That also works, but will often force callers to implement extra `null` checks. The contract is that the return value would be `null` only when the input argument `url` is `null`. To express that contract, you would annotate this method as shown in the following code:
+
+```csharp
+[return: NotNullWhenNotNull("url")]
+string? GetTopLevelDomainFromFullUrl(string? url);
+```
+
+The return value and the argument have both been annotated with the `?` indicating that either could be `null`. The attribute further clarifies that the return value won't be null when the `url` argument isn't `null`. You specify conditional postconditions using these attributes:
+
+- [MaybeNullWhen](xref:System.Diagnostics.CodeAnalysis.MaybeNullWhenAttribute): A non-nullable `out` or `ref` argument may be null when the return value satisfies a condition.
+- [NotNullWhen](xref:System.Diagnostics.CodeAnalysis.NotNullWhenAttribute): A nullable `out` or `ref` argument may not be null when the return value satisfies a condition.
+- [NotNullIfNotNull](xref:System.Diagnostics.CodeAnalysis.NotNullIfNotNullAttribute): a string return value isn't null when the input string argument isn't null.
+
+## Generic definitions and nullability
+
+Correctly communicating the null state of generic types and generic methods requires special care. This stems from the fact that a nullable value type and a nullable reference type are fundamentally different. An `int?` is a synonym for `Nullable<int>`, whereas `string?` is `string` with an attribute added by the compiler. The result is that the compiler can't generate correct code for `T?` without knowing if `T` is a `class` or a `struct`. 
+
+This doesn't mean you can use a nullable type (either value type or reference type) as the type argument for a closed generic type. Both `List<string?>` and `List<int?>` are valid instantiations of `List<T>`. 
+
+What it does mean is that you can't use `T?` in a generic class or method declaration without constraints. For example, <xref:System.Linq.Enumerable%601(System.Collections.Generic.IEnumerable%601)?displayProperty=nameWithType> won't be changed to return `T?`. You can overcome this limitation by adding either the `struct` or `class` constraint. With either of those constraints, the compiler knows how to generate code for both `T` and `T?`.
```

Comment:
The link here is broken, and I'm not sure of the correct syntax for a closed generic type. Perhaps do it as `[Enumerable<IEnumerable>](xref:System.collectionsGeneric.IEnumerable?601)`?

Reply:
trying one other technique. (I don't know either)